sessionctx/variable/session.go

```diff
@@ -1907,3 +1927,130 @@ func (s *SessionVars) SlowLogFormat(logItems *SlowQueryLogItems) string {
 func writeSlowLogItem(buf *bytes.Buffer, key, value string) {
 	buf.WriteString(SlowLogRowPrefixStr + key + SlowLogSpaceMarkStr + value + "\n")
 }
+
+// ResetTxnTimes reset the record times for session.
+func (s *SessionVars) ResetTxnTimes() {
+	s.StartTimes = s.StartTimes[:0]
+	s.EndTimes = s.EndTimes[:0]
+}
+
+// ReportTxnAndLockDuration reports txn and the pessimistic locks' idle and duration to metric server.
+// Run this function in another goroutine to avoid it blocks the Commit or Rollback statements.
+func (s *SessionVars) ReportTxnAndLockDuration(startTS uint64, keysArr []kv.LockedKeys) {
+	if len(s.EndTimes) == 0 {
+		return
+	}
+	// copy the slice because it may be cleaned in other goroutines
+	startTimes := make([]time.Time, len(s.StartTimes))
+	copy(startTimes, s.StartTimes)
+	endTimes := make([]time.Time, len(s.EndTimes))
+	copy(endTimes, s.EndTimes)
+	// the rest works can be executed in another goroutine, so that it won't block the session goroutine.
+	go func() {
+		lockedKeysDurations, txnRes, err := CalcLockDurations(keysArr, startTimes, endTimes)
+		if err != nil {
+			logutil.BgLogger().Info("[long-txn] calc lock duration error", zap.Error(err))
+			return
+		}
+
+		// observe the lock duration of the transaction
+		sessionTxnObtainLockIdleDuration.Observe(txnRes.LockIdleDuration.Seconds())
+		sessionTxnObtainLockExecDuration.Observe(txnRes.LockExecDuration.Seconds())
+		// observe the total duration of the transaction
+		sessionTxnNoneLockIdleDuration.Observe(txnRes.TotalIdleDuration.Seconds())
+		sessionTxnNoneLockExecDuration.Observe(txnRes.TotalExecDuration.Seconds())
+		// observe durations for each lock
+		for _, locksDuration := range lockedKeysDurations {
+			for i := 0; i < locksDuration.Keys; i++ {
+				sessionLockIdleDuration.Observe(locksDuration.IdleDuration.Seconds())
+				sessionLockExecDuration.Observe(locksDuration.ExecDuration.Seconds())
+			}
+		}
+
+		threshold := time.Duration(atomic.LoadInt64(&LongTxnThreshold)) * time.Millisecond
+		if startTimes[len(startTimes)-1].Sub(endTimes[0]) < threshold {
```

Comment:
Why we check the threshold by last start time - first end time instead of last end time(now?) - first start time?